File: Data Structures & Algorithms/longest-consecutive-sequence/submission-1.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        // O(n log n)
        sort(nums.begin(), nums.end());

        int ans = 1;
        int max_ans = ans;

        for(int i=1; i<nums.size(); i++) {
            if(nums[i-1] == nums[i]) 
                continue;
            if(nums[i-1] == nums[i]-1)
                ans++;
            else {
                if(ans > max_ans) {
                    max_ans = ans;
                }
                ans = 1;
            }
        }
        if(ans > max_ans) {
            max_ans = ans;
        }
        return max_ans;
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.java
class Solution {
    private Set<List<Integer>> ans = new HashSet<>();
    int target;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        this.target = target;
        Arrays.sort(candidates);
        addSubsets(new ArrayList<>(), candidates, 0, 0);
        return ans.stream().collect(Collectors.toList());
    }
    public void addSubsets(ArrayList<Integer> already, int[] nums, int i, int sum) {
        for(;i<nums.length;i++) {
            List<Integer> temp = new ArrayList<>(already);
            temp.add(nums[i]);
            sum+=nums[i];
            if(sum==target)
                ans.add(temp);
            // ans.add((new ArrayList<>(already)).add(nums[i]));
            already.add(nums[i]);
            addSubsets(already, nums, i+1, sum);
            already.remove(already.size()-1);
            sum-=nums[i];
        }
    }
}

class Solution_Previous_Question {
    private List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        ans.add(new ArrayList<>());
        addSubsets(new ArrayList<>(), nums, 0);
        return ans;
    }
    public void addSubsets(ArrayList<Integer> already, int[] nums, int i) {
        for(;i<nums.length;i++) {
            List<Integer> temp = new ArrayList<>(already);
            temp.add(nums[i]);
            ans.add(temp);
            // ans.add((new ArrayList<>(already)).add(nums[i]));
            already.add(nums[i]);
            addSubsets(already, nums, i+1);
            already.remove(already.size()-1);
        }
    }
}